### ISSUE TITLE
Add a python API

### ISSUE THREAD

**Issue by edwardhartnett** [NOAA-EMC/NCEPLIBS-g2c#187]
Add a python API.

**Comment by kgerheiser** [comment]
To call this library from Python we'll need to make a shared library build.

**Comment by DusanJovic-NOAA** [comment]
Why does python API need to be added to this library? There are already many different python APIs for decoding/encoding GRIB2 files. For example: https://github.com/jswhit/pygrib or https://github.com/eengl/grib2io or https://www.cpc.ncep.noaa.gov/products/wesley/wgrib2/pywgrib2_s.html. How will this API be different than all of these.
I think this library should be just a C implementation of grib2 decoder/encoder, nothing else. Bindings to other languages should be in different libraries/projects.

**Comment by edwardhartnett** [comment]
Unfortunately neither of those python libraries are suitable for all our needs. The first because it uses ecCodes, which is not approved for operational machines, and the second depends on wgrib2, which we are trying to move away from due to repeated difficulties building and installing.

Thus a new python API is called for. 

This library is intended to grow in any case: we are going to extend its features and add a command line tool. We will keep everything in this repo because we already have too many repos. ;-) But we will allow a non-python build with CMake options.

**Comment by DusanJovic-NOAA** [comment]
pygrib used to be wrapper around g2clib, which they removed and replaced it with eccodes. I should have checked that, so sorry for listing it as an example. grib2io is definitely using g2c it even says that in the description, where do you see that it depends on wgrib2. wgrib2 itself depends (optionally) on g2c, which unfortunately is using very old version (1.4.0) due to some bugs in g2c which we should try to fix. wgrib2 command line tool is de-facto standard tool for processing grib2 files, used in many production systems. If we had difficulties building it we should work on fixing those instead of inventing completely new set of tools and expecting everybody to switch to these new tools. Just my opinion.
I still think we should keep this library (and a repo) small, standalone decoding/encoding library without any tools or APIs in different languages. Bindings to different languages should be kept elsewhere (in different repositories).

**Comment by edwardhartnett** [comment]
We have put a lot of effort and time into improving wgrib2, but without the cooperation of the developer, we get nowhere.

I didn't know about grib2io, I will take a look...

**Comment by edwardhartnett** [comment]
grib2io is the official Python API, so I will close this issue.

I would love it if wgrib2 could be built and maintained reasonably, but it's a giant mess. The code is not even in a repo, the build system is a home-rolled nightmare which includes actual COPIES of old netCDF and other libraries, and there is no unit testing, making modifications impossible to verify.

Kyle spent the time to develop a cmake build system, but the maintainer of the software refused to use it and just carried on making changes in his home-rolled hillbilly build system, breaking the CMake build for us.

Basically wgrib2 is a great example of how hundreds of thousands of dollars can be spent on software, and we still end up with something so terrible we have to spend extra money tearing it out and replacing it. ;-)

We have tried to work with the author but he just ignores us and our needs. It's hopeless.
